virttest/migration.py

```diff
@@ -578,6 +579,29 @@ def set_migratepostcopy(self, vm_name, uri=None):
             raise exceptions.TestError("Unalbe to find event {}"
                                        .format(exp_str))
 
+    def control_migrate_speed(self, vm_name, to_speed=1, opts=""):
+        """
+        Set migration speed to control the migration duration
+
+        :param vm_name: vm name
+        :param to_speed: the speed value in Mbps to be set for migration
+        :param opts: the options for virsh command
+        """
+
+        def _set_speed(vm_name, to_speed, opts=''):
+            virsh_args = {}
+            virsh_args.update({"ignore_status": False})
+            old_speed = virsh.migrate_getspeed(vm_name, extra=opts, **virsh_args)
+            logging.debug("Current %s migration speed is %s MiB/s\n", opts, old_speed.stdout_text.strip())
+            logging.debug("Set %s migration speed to %d MiB/s\n", opts, to_speed)
+            virsh.migrate_setspeed(vm_name, to_speed, extra=opts, **virsh_args)
+            actual_speed = virsh.migrate_getspeed(vm_name, extra=opts, **virsh_args)
+            logging.debug("New %s migration speed is %s MiB/s\n", opts, actual_speed.stdout_text.strip())
+
+        _set_speed(vm_name, to_speed)
+        if '--postcopy' in opts and libvirt_version.version_compare(5, 0, 0):
+            _set_speed(vm_name, to_speed, opts=opts)
```

Comment:
I can imagine why you set the speed for both precopy and postcopy here, we may need to slowdown the whole process. But I'm still feel a little bit confused that when **opts=='--postcopy'**, it setsspeed for precopy as well. is it expected? I'm wondering if we need a parameter(eg, mode) to specify whether to set pre-cpy, postcopy or both precopy-cpy and postcopy migration speed. What do you think?

Reply:
When opts=='--postcopy', it setsspeed for precopy as well. This is usually what feature owner does in their test. So I would like to mock up this step to ensure the migration has enough time to shift to postcopy from precopy.

Reply:
Ok, I see. But I think we'd better set a switch here... or move _set_speed to outside of the control_migrate_speed, in case we only need to set postcopy speed.

Reply:
Thanks for your suggestion. I think it is a good idea. Updated.